pinot-core/src/main/java/org/apache/pinot/core/plan/SelectionPlanNode.java

```diff
@@ -32,66 +30,65 @@
 import org.apache.pinot.core.operator.query.SelectionOnlyOperator;
 import org.apache.pinot.core.operator.query.SelectionOrderByOperator;
 import org.apache.pinot.core.operator.transform.TransformOperator;
-import org.apache.pinot.pql.parsers.pql2.ast.IdentifierAstNode;
+import org.apache.pinot.core.query.request.context.ExpressionContext;
+import org.apache.pinot.core.query.request.context.OrderByExpressionContext;
+import org.apache.pinot.core.query.request.context.QueryContext;
+import org.apache.pinot.core.query.selection.SelectionOperatorUtils;
 
 
 /**
  * The <code>SelectionPlanNode</code> class provides the execution plan for selection query on a single segment.
  */
 public class SelectionPlanNode implements PlanNode {
   private final IndexSegment _indexSegment;
-  private final Selection _selection;
+  private final QueryContext _queryContext;
   private final TransformPlanNode _transformPlanNode;
 
-  public SelectionPlanNode(IndexSegment indexSegment, BrokerRequest brokerRequest) {
+  public SelectionPlanNode(IndexSegment indexSegment, QueryContext queryContext) {
     _indexSegment = indexSegment;
-    _selection = brokerRequest.getSelections();
-    _transformPlanNode =
-        new TransformPlanNode(_indexSegment, brokerRequest, collectExpressionsToTransform(indexSegment, brokerRequest));
+    _queryContext = queryContext;
+    _transformPlanNode = new TransformPlanNode(_indexSegment, queryContext, collectExpressions());
   }
 
   @Override
   public Operator<IntermediateResultsBlock> run() {
     TransformOperator transformOperator = _transformPlanNode.run();
-    if (_selection.getSize() > 0) {
-      if (_selection.getSelectionSortSequence() == null) {
-        return new SelectionOnlyOperator(_indexSegment, _selection, transformOperator);
+    Selection selection = _queryContext.getBrokerRequest().getSelections();
+    if (_queryContext.getLimit() > 0) {
+      if (_queryContext.getOrderByExpressions() == null) {
+        return new SelectionOnlyOperator(_indexSegment, selection, transformOperator);
       } else {
-        return new SelectionOrderByOperator(_indexSegment, _selection, transformOperator);
+        return new SelectionOrderByOperator(_indexSegment, selection, transformOperator);
       }
     } else {
-      return new EmptySelectionOperator(_indexSegment, _selection, transformOperator);
+      return new EmptySelectionOperator(_indexSegment, selection, transformOperator);
     }
   }
 
-  private Set<TransformExpressionTree> collectExpressionsToTransform(IndexSegment indexSegment,
-      BrokerRequest brokerRequest) {
-
-    Set<TransformExpressionTree> expressionTrees = new LinkedHashSet<>();
-    Selection selection = brokerRequest.getSelections();
+  private Set<TransformExpressionTree> collectExpressions() {
+    Set<TransformExpressionTree> expressionTrees = new HashSet<>();
```

Comment:
I think we should add a TODO in this and related places that eventually (in the end state of this migration), this compilation to TransformExpressionTree will not be needed.

Reply:
This will be cleaned up very soon, so I think the TODO is not necessary (or there will be too many TODOs added).
BTW, the `toTransformExpressionTree()` does not require any compilation, so the per-segment compilation for selection is already eliminated.